File: tools/build_rom.py
#!/usr/bin/env python3
"""Link the full ROM from tracked source with a GNU linker script (pret model).

Unlike build_full.py, which overlays claimed regions onto a copy of the ROM,
this assembles the whole image the way pret decompilations do: every claimed C
and assembly function is compiled/assembled to an object placed at its fixed
address, every claimed asset is the byte output of its source encoder, and every
not-yet-reconstructed gap is `.incbin "baserom.gba", offset, size` (the skeleton
that shrinks as regions are claimed). One `arm-none-eabi-ld` invocation produces
the ELF; objcopy produces the .gba; the result is checked byte-identical.

Requires build_claimed.py, build_asm.py and build_assets.py to have populated
out/claimed, out/asm and out/assets first (this driver runs them).
"""
import argparse
import json
import subprocess
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
ROM_BASE = 0x08000000
EXTERNAL = __import__("re").compile(r"(Func|Data|Value)_([0-9a-f]{8})")


def run(command):
    result = subprocess.run(command, cwd=ROOT, text=True,
                            capture_output=True)
    if result.returncode:
        raise RuntimeError((result.stderr or result.stdout).strip())
    return result


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("rom", type=Path, nargs="?", default="baserom.gba")
    parser.add_argument("-o", "--output", type=Path, default=ROOT / "out/rom")
    parser.add_argument("--jobs", type=int, default=8)
    args = parser.parse_args()
    rom = args.rom.read_bytes()
    out = args.output
    out.mkdir(parents=True, exist_ok=True)

    # 三種のソース群を構築し、それぞれ正確な領域一覧を出力する。
    run(["python3", "tools/build_claimed.py", str(args.rom),
         "--jobs", str(args.jobs), "--output", "out/claimed"])
    run(["python3", "tools/build_asm.py", str(args.rom), "--output", "out/asm"])
    run(["python3", "tools/build_assets.py", str(args.rom), "--output", "out/assets"])

    # 各確定領域は検証済みの正確なバイト列を供給する。オブジェクトは
    # ソースからコンパイル・アセンブルし、生成バイトを
    # build_claimed/build_asmでROMと照合する。Cコンパイラは
    # ROM作成時のarmccとは埋込み方が異なるため、断片は正確な
    # サイズのincbinとして置く。build_bytes/には
    # リンカスクリプトがアドレス順に置く領域別バイナリを格納する。
    bytes_dir = out / "regions"
    bytes_dir.mkdir(exist_ok=True)
    pieces = []  # （アドレス、サイズ、種別、内容）
    claimed = json.loads((ROOT / "out/claimed/manifest.json").read_text())
    image = (ROOT / "out/claimed/claimed.bin").read_bytes()
    image_base = claimed["image_base"]
    for region in claimed["regions"]:
        start = region["address"] - image_base
        blob = bytes_dir / f"{region['address']:08x}.bin"
        blob.write_bytes(image[start:start + region["size"]])
        pieces.append((region["address"], region["size"], "incbin",
                       str(blob.relative_to(ROOT))))
    asm_manifest = json.loads((ROOT / "out/asm/manifest.json").read_text())
    for region in asm_manifest["regions"]:
        pieces.append((region["address"], region["size"], "incbin",
                       str(Path(region["output"]).relative_to(ROOT))))
    assets = json.loads((ROOT / "out/assets/manifest.json").read_text())
    for region in assets["regions"]:
        pieces.append((region["address"], region["size"], "incbin",
                       str(Path(region["output"]).relative_to(ROOT))))

    pieces.sort()
    # 確定済み断片間の全間隙をbaseromのincbin骨格で埋める。
    filled = []
    cursor = ROM_BASE
    for address, size, kind, payload in pieces:
        if address < cursor:
            raise SystemExit(f"overlapping piece at 0x{address:08x}")
        if address > cursor:
            filled.append((cursor, address - cursor, "skeleton", None))
        filled.append(（アドレス、サイズ、種別、内容）)
        cursor = address + size
    if cursor < ROM_BASE + len(rom):
        filled.append((cursor, ROM_BASE + len(rom) - cursor, "skeleton", None))

    # 一つのオブジェクトに各領域を個別セクションとして格納する。確定バイトと
    # 素材バイトは検証済み領域バイナリの.incbin、骨格間隙は
    # baserom.gbaのoffset・size指定incbinで格納する（pret式incbin骨格）。
    fill_lines = [".syntax unified"]
    placements = []
    for address, size, kind, payload in filled:
        name = f".rom_{address:08x}"
        fill_lines.append(f'\t.section {name}, "ax", %progbits')
        if kind == "incbin":
            fill_lines.append(f'\t.incbin "{payload}"')
        else:
            fill_lines.append(
                f'\t.incbin "{args.rom}", 0x{address - ROM_BASE:x}, 0x{size:x}')
        placements.append(f"    KEEP(out/rom/rom_fill.o({name}))")
    (out / "rom_fill.s").write_text("\n".join(fill_lines) + "\n")
    run(["arm-none-eabi-as", "-mcpu=arm7tdmi", "-mthumb-interwork",
         "-o", str(out / "rom_fill.o"), str(out / "rom_fill.s")])

    # リンカスクリプトでは全領域のセクションをアドレス順に置き、ROMを
    # 0x08000000から連続させる。
    (out / "ld_script.ld").write_text(
        "OUTPUT_ARCH(arm)\n"
        "MEMORY { ROM (rx) : ORIGIN = 0x08000000, LENGTH = 32M }\n"
        "SECTIONS\n{\n"
        "  .rom 0x08000000 : SUBALIGN(1)\n  {\n" + "\n".join(placements) +
        "\n  } > ROM\n"
        "  /DISCARD/ : { *(.comment) *(.note*) *(.ARM.*) }\n"
        "}\n")

    elf = out / "rom.elf"
    run(["arm-none-eabi-ld", "-T", str(out / "ld_script.ld"),
         "-o", str(elf), str((out / "rom_fill.o").relative_to(ROOT))])
    built = out / "rom.gba"
    run(["arm-none-eabi-objcopy", "-O", "binary", str(elf), str(built)])
    image = built.read_bytes()
    if image != rom:
        first = next((i for i in range(min(len(image), len(rom)))
                      if image[i] != rom[i]), None)
        raise SystemExit(
            f"linked ROM differs (size {len(image):#x} vs {len(rom):#x}, "
            f"first diff at 0x{ROM_BASE + first:08x})" if first is not None
            else f"linked ROM size {len(image):#x} vs {len(rom):#x}")
    claimed_bytes = sum(s for _, s, k, _ in filled if k != "skeleton")
    regions = sum(1 for _, _, k, _ in filled if k != "skeleton")
    print(f"identical=True pieces={len(filled)} regions={regions} "
          f"source_bytes={claimed_bytes} "
          f"skeleton_bytes={len(rom) - claimed_bytes}")


if __name__ == "__main__":
    main()

File: tools/build_rom.ts
#!/usr/bin/env bun
// Tool role: entrypoint; invoked by package.json.
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { basename, dirname, isAbsolute, join, relative, resolve } from "node:path";

const ROOT = dirname(dirname(Bun.fileURLToPath(import.meta.url)));
const ROM_BASE = 0x08000000;

type PieceKind = "incbin" | "skeleton";
type Piece = [address: number, size: number, kind: PieceKind, payload: string | null];

interface Options {
  rom: string;
  output: string;
  jobs: number;
}

function rooted(path: string): string {
  return isAbsolute(path) ? path : resolve(ROOT, path);
}

function usage(): void {
  console.log("usage: build_rom.ts [-h] [-o OUTPUT] [--jobs JOBS] [rom]");
}

function parseArgs(argv: string[]): Options {
  const options: Options = { rom: "roms/gs1-en.gba", output: "out/rom", jobs: 8 };
  let positional = false;
  for (let index = 0; index < argv.length; index++) {
    const argument = argv[index];
    if (argument === "-h" || argument === "--help") {
      usage();
      process.exit(0);
    }
    const equal = argument.indexOf("=");
    const option = equal >= 0 ? argument.slice(0, equal) : argument;
    if (option === "-o" || option === "--output" || option === "--jobs") {
      const value = equal >= 0 ? argument.slice(equal + 1) : argv[++index];
      if (value === undefined) throw new Error(`${option} requires a value`);
      if (option === "--jobs") options.jobs = Number.parseInt(value, 10);
      else options.output = value;
    } else if (!argument.startsWith("-") && !positional) {
      options.rom = argument;
      positional = true;
    } else {
      throw new Error(`unrecognized argument: ${argument}`);
    }
  }
  if (!Number.isInteger(options.jobs) || options.jobs < 1) throw new Error("jobs must be positive");
  return options;
}

async function run(command: string[]): Promise<string> {
  const child = Bun.spawn(command, { cwd: ROOT, stdout: "pipe", stderr: "pipe" });
  const [stdout, stderr, exitCode] = await Promise.all([
    new Response(child.stdout).text(),
    new Response(child.stderr).text(),
    child.exited,
  ]);
  if (exitCode !== 0) throw new Error((stderr || stdout).trim() || `${basename(command[0])} failed`);
  return stdout;
}

function readJson(path: string): any {
  return JSON.parse(readFileSync(path, "utf8"));
}

async function main(): Promise<void> {
  const args = parseArgs(Bun.argv.slice(2));
  const romPath = resolve(process.cwd(), args.rom);
  const rom = readFileSync(romPath);
  const out = rooted(args.output);
  mkdirSync(out, { recursive: true });

  // 三種のソース群を構築し、それぞれ正確な領域一覧を出力する。
  await run([process.execPath, "tools/build_claimed.ts", romPath,
    "--jobs", String(args.jobs), "--output", "out/claimed"]);
  await run([process.execPath, "tools/build_asm.ts", romPath, "--output", "out/asm"]);
  await run([process.execPath, "tools/build_assets.ts", romPath, "--output", "out/assets"]);

  // 各確定領域は検証済みの正確なバイト列を供給する。
  // ソースから構築した領域別バイナリをアドレス順に配置する。
  const bytesDir = join(out, "regions");
  mkdirSync(bytesDir, { recursive: true });
  const pieces: Piece[] = [];
  const claimed = readJson(join(ROOT, "out/claimed/manifest.json"));
  const image = readFileSync(join(ROOT, "out/claimed/claimed.bin"));
  const imageBase = Number(claimed.image_base);
  for (const region of claimed.regions) {
    const address = Number(region.address);
    const size = Number(region.size);
    const start = address - imageBase;
    const blob = join(bytesDir, `${address.toString(16).padStart(8, "0")}.bin`);
    writeFileSync(blob, image.subarray(start, start + size));
    pieces.push([address, size, "incbin", relative(ROOT, blob)]);
  }
  const asmManifest = readJson(join(ROOT, "out/asm/manifest.json"));
  for (const region of asmManifest.regions) {
    pieces.push([
      Number(region.address), Number(region.size), "incbin",
      relative(ROOT, region.output),
    ]);
  }
  const assets = readJson(join(ROOT, "out/assets/manifest.json"));
  for (const region of assets.regions) {
    pieces.push([
      Number(region.address), Number(region.size), "incbin",
      relative(ROOT, region.output),
    ]);
  }

  pieces.sort((left, right) => left[0] - right[0] || left[1] - right[1]);
  // 確定済み断片間の全間隙をbaseromのincbin骨格で埋める。
  const filled: Piece[] = [];
  let cursor = ROM_BASE;
  for (const [address, size, kind, payload] of pieces) {
    if (address < cursor) {
      throw new Error(`overlapping piece at 0x${address.toString(16).padStart(8, "0")}`);
    }
    if (address > cursor) filled.push([cursor, address - cursor, "skeleton", null]);
    filled.push([address, size, kind, payload]);
    cursor = address + size;
  }
  if (cursor < ROM_BASE + rom.length) {
    filled.push([cursor, ROM_BASE + rom.length - cursor, "skeleton", null]);
  }

  // 一つのオブジェクトに各領域を個別セクションとして格納する。
  const fillLines = [".syntax unified"];
  const placements: string[] = [];
  const fillObject = join(out, "rom_fill.o");
  for (const [address, size, kind, payload] of filled) {
    const name = `.rom_${address.toString(16).padStart(8, "0")}`;
    fillLines.push(`\t.section ${name}, "ax", %progbits`);
    if (kind === "incbin") fillLines.push(`\t.incbin "${payload}"`);
    else fillLines.push(`\t.incbin "${romPath}", 0x${(address - ROM_BASE).toString(16)}, 0x${size.toString(16)}`);
    placements.push(`    KEEP(${relative(ROOT, fillObject)}(${name}))`);
  }
  const fillSource = join(out, "rom_fill.s");
  writeFileSync(fillSource, fillLines.join("\n") + "\n");
  await run([
    "arm-none-eabi-as", "-mcpu=arm7tdmi", "-mthumb-interwork",
    "-o", fillObject, fillSource,
  ]);

  // 全領域を0x08000000からアドレス順に連続配置する。
  const linker = join(out, "ld_script.ld");
  writeFileSync(linker,
    "OUTPUT_ARCH(arm)\n" +
    "MEMORY { ROM (rx) : ORIGIN = 0x08000000, LENGTH = 32M }\n" +
    "SECTIONS\n{\n" +
    "  .rom 0x08000000 : SUBALIGN(1)\n  {\n" + placements.join("\n") +
    "\n  } > ROM\n" +
    "  /DISCARD/ : { *(.comment) *(.note*) *(.ARM.*) }\n" +
    "}\n",
  );
  const elf = join(out, "rom.elf");
  await run(["arm-none-eabi-ld", "-T", linker, "-o", elf, relative(ROOT, fillObject)]);
  const built = join(out, "rom.gba");
  await run(["arm-none-eabi-objcopy", "-O", "binary", elf, built]);
  const builtImage = readFileSync(built);
  if (!builtImage.equals(rom)) {
    let first: number | undefined;
    const common = Math.min(builtImage.length, rom.length);
    for (let index = 0; index < common; index++) {
      if (builtImage[index] !== rom[index]) {
        first = index;
        break;
      }
    }
    if (first !== undefined) {
      throw new Error(
        `linked ROM differs (size 0x${builtImage.length.toString(16)} vs 0x${rom.length.toString(16)}, ` +
        `first diff at 0x${(ROM_BASE + first).toString(16).padStart(8, "0")})`,
      );
    }
    throw new Error(`linked ROM size 0x${builtImage.length.toString(16)} vs 0x${rom.length.toString(16)}`);
  }
  const claimedBytes = filled
    .filter(([, , kind]) => kind !== "skeleton")
    .reduce((sum, [, size]) => sum + size, 0);
  const regions = filled.filter(([, , kind]) => kind !== "skeleton").length;
  console.log(
    `identical=True pieces=${filled.length} regions=${regions} source_bytes=${claimedBytes} ` +
    `skeleton_bytes=${rom.length - claimedBytes}`,
  );
}

if (import.meta.main) await main();
